crates/net/downloaders/src/file_codec_ovm_receipt.rs

```diff
@@ -0,0 +1,286 @@
+//! Codec for reading raw receipts from a file.
+
+use alloy_rlp::{Decodable, RlpDecodable, RlpDecodableWrapper};
+use reth_primitives::{
+    bytes::{Buf, BytesMut},
+    Address, Bloom, Bytes, Log, Receipt, TxType, B256,
+};
+use tokio_util::codec::Decoder;
+
+use crate::{file_client::FileClientError, receipt_file_client::ReceiptWithBlockNumber};
+
+/// Codec for reading raw receipts from a file.
+///
+/// If using with [`FramedRead`](tokio_util::codec::FramedRead), the user should make sure the
+/// framed reader has capacity for the entire receipts file. Otherwise, the decoder will return
+/// [`InputTooShort`](alloy_rlp::Error::InputTooShort), because RLP receipts can only be
+/// decoded if the internal buffer is large enough to contain the entire receipt.
+///
+/// Without ensuring the framed reader has capacity for the entire file, a receipt is likely to
+/// fall across two read buffers, the decoder will not be able to decode the receipt, which will
+/// cause it to fail.
+///
+/// It's recommended to use [`with_capacity`](tokio_util::codec::FramedRead::with_capacity) to set
+/// the capacity of the framed reader to the size of the file.
+#[derive(Debug)]
+pub struct ReceiptFileCodec;
+
+impl Decoder for ReceiptFileCodec {
+    type Item = ReceiptWithBlockNumber;
+    type Error = FileClientError;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        if src.is_empty() || src.len() == 1 {
+            return Ok(None)
+        }
```

Comment:
found in testing:
```suggestion
        if src.is_empty() {
            return Ok(None)
        }
```

this is incorrect because when `FramedRead` calls `decode_eof`, we will return with `Ok(None)`, and a non-empty `src`.

See:
https://docs.rs/tokio-util/latest/src/tokio_util/codec/decoder.rs.html#151

This means we must only return `Ok(None)` if `src.is_empty()`.

Reply:
ah, realizing we have to keep this condition, so we probably just advance the buf if len == 1

Reply:
this is a hack because `HackReceipt` encodes as `HackReceiptContainer { HackReceipt { .. }}`. so the genesis block gets a mapping of `HackReceiptContainer { empty }` and so does the bedrock block - those two blocks have no transactions. 
```rust
#[rlp(trailing)]
HackReceiptContainer {
    receipt: Option<HackReceipt>
}
```
should achieve this too right? and then checking if `HackReceipt.receipt.is_none()` -> decode next receipt without returning in-between

Reply:
> should achieve this too right? and then checking if HackReceipt.receipt.is_none() -> decode next receipt without returning in-between

Yeah I think that should work w.r.t. decoding

Reply:
did you already impl this neatly? otherwise I'll do it

Reply:
> did you already impl this neatly? otherwise I'll do it

nope, did not impl this nicely yet